packages/react-components/react-dialog/Spec.md

```diff
@@ -1,63 +1,459 @@
+  content="Connect driver"
+  header="Transmit capacitor"
+  headerAction="Generate protocol"
+  trigger={<Button content="A trigger" />}
+/>
+```
 
 ## API
 
-_List the **Props** and **Slots** proposed for the component. Ideally this would just be a link to the component's `.types.ts` file_
+The `Dialog` should implement a `children` based API as is one of the standards across all the surveyed alternatives as a part of Open UI research in [Prior Art](#prior-art). The component will leverage the use of `context` in the interaction and data flows of child compound components.
 
-## Structure
+Sample usages will be give in the following section of this document [Sample code](#sample-code)
 
-- _**Public**_
-- _**Internal**_
-- _**DOM** - how the component will be rendered as HTML elements_
+### Dialog
 
-## Migration
+The root level component serves as an interface for interaction with all possible behaviors exposed. It provides context down the hierarchy to `children` compound components to allow functionality.
+
+```tsx
+type DialogSlots = {
+  /**
+   * The dialog element itself
+   */
+  root: Slot<'div'>;
+  /**
+   * Dimmed background of dialog.
+   * The default overlay is rendered as a `<div>` with styling.
+   * This slot expects a `<div>` element which will replace the default overlay.
+   * The overlay should have `aria-hidden="true"`.
+   */
+  overlay?: Slot<'div'>;
+};
+
+type DialogProps = ComponentProps<DialogSlots> & {
+  /**
+   * Dialog variations.
+   *
+   * `modal`: When this type of dialog is open, the rest of the page is dimmed out and cannot be interacted with. The tab sequence is kept within the dialog and moving the focus outside the dialog will imply closing it. This is the default type of the component.
+   *
+   * `non-modal`: When a non-modal dialog is open, the rest of the page is not dimmed out and users can interact with the rest of the page. This also implies that the tab focus can move outside the dialog when it reaches the last focusable element.
+   *
+   * `alert`: is a special type of modal dialogs that interrupts the user's workflow to communicate an important message or ask for a decision. Unlike a typical modal dialog, the user must take an action through the options given to dismiss the dialog, and it cannot be dismissed through the dimmed background or escape key.
+   */
+  type?: 'modal' | 'non-modal' | 'alert';
+  /**
+   * Controls the open state of the dialog
+   * @default undefined
+   */
+  open?: boolean;
+  /**
+   * Default value for the uncontrolled open state of the dialog.
+   * @default false
+   */
+  defaultOpen?: boolean;
+  /**
+   * Callback fired when the component requests to be closed.
+   * @default undefined
+   */
+  onClose?(event: MouseEvent | KeyboardEvent, data: DialogCloseData): void;
+};
+
+type DialogCloseData = {
+  /**
+   * The event source of the callback invocation
+   */
+  type: 'escapeKeyDown' | 'overlayClick' | 'closeButtonClick';
```

Comment:
this might be dangerous - if consumers would only check `closeButtonClick` (because they usually think about mouse users). It would be good to have the intent as a primary indicator in the api - so type would be confirm or cancel. We can have some less obvious prop for details (keyboard, button click, outside click)

Reply:
`confirm` and `cancel` aren't states relevant for the logic of a dialog per se. I'm not sure we should bind this intent idea as it only is associated with formularies scenarios and that could be achieved by simply controlling the component instead.

Reply:
Also, `closeButtonClick` is activated on `enter` and `space` click as it'll be a button, so even by only checking `closeButtonClick` users will be veryfying more than just mouse click.